File: golang-devildyw/simple/L121买卖股票的最佳时机.go
package main

import "math"

func main() {
	maxProfit([]int{7, 1, 5, 3, 6, 4})
}

func maxProfit(prices []int) int {
	// 从左往右遍历，更新过程中的买入最小值 并不断更新利润 初始化买入最小值是第一天
	var (
		minPrice = math.MaxInt32
		result   = 0
	)

	for i := 0; i < len(prices); i++ {
		if minPrice > prices[i] {
			minPrice = prices[i]
		} else {
			profit := prices[i] - minPrice
			if profit > result {
				result = profit
			}
		}
	}
	return result
}

File: java-devildyw/src/main/java/top/dyw/leetcode/Hot100/L121买卖股票的最佳时机.java
package top.dyw.leetcode.Hot100;

public class L121买卖股票的最佳时机 {
    public int maxProfit(int[] prices) {
        int profie = prices[0];
        int result = 0;
        for (int i=1; i<prices.length; i++) {
            result = Math.max(result, prices[i] - profie);
            profie = Math.min(profie, prices[i]);
        }
        return result;
    }
}
